model.py

```diff
@@ -10320,6 +10326,89 @@ def with_password(cls, _db):
         """Get Admins that have a password."""
         return _db.query(Admin).filter(Admin.password_hashed != None)
 
+    def is_system_admin(self):
+        _db = Session.object_session(self)
+        role = get_one(_db, AdminRole, admin=self, role=AdminRole.SYSTEM_ADMIN)
+        if role:
+            return True
+        return False
+
+    def is_sitewide_library_manager(self):
+        _db = Session.object_session(self)
+        if self.is_system_admin():
+            return True
+        role = get_one(_db, AdminRole, admin=self, role=AdminRole.SITEWIDE_LIBRARY_MANAGER)
+        if role:
+            return True
+        return False
+
+    def is_sitewide_librarian(self):
+        _db = Session.object_session(self)
+        if self.is_sitewide_library_manager():
+            return True
+        role = get_one(_db, AdminRole, admin=self, role=AdminRole.SITEWIDE_LIBRARIAN)
+        if role:
+            return True
+        return False
+
+    def is_library_manager(self, library):
+        _db = Session.object_session(self)
+        # First check if the admin is a manager of _all_ libraries.
+        if self.is_sitewide_library_manager():
+            return True
+        # If not, they could stil be a manager of _this_ library.
+        role = get_one(_db, AdminRole, admin=self, library=library, role=AdminRole.LIBRARY_MANAGER)
+        if role:
+            return True
+        return False
+
+    def is_librarian(self, library):
+        _db = Session.object_session(self)
+        # If the admin is a library manager, they can do everything a librarian can do.
+        if self.is_library_manager(library):
+            return True
+        # Check if the admin is a librarian for _all_ libraries.
+        if self.is_sitewide_librarian():
+            return True
+        # If not, they might be a librarian of _this_ library.
+        role = get_one(_db, AdminRole, admin=self, library=library, role=AdminRole.LIBRARIAN)
+        if role:
+            return True
+        return False
+
+    def add_role(self, role, library=None):
+        _db = Session.object_session(self)
+        role, is_new = get_one_or_create(_db, AdminRole, admin=self, role=role, library=library)
+        return role
+
+    def remove_role(self, role, library=None):
+        _db = Session.object_session(self)
+        role = get_one(_db, AdminRole, admin=self, role=role, library=library)
+        if role:
+            _db.delete(role)
+
+class AdminRole(Base):
```

Comment:
I was thinking this should subclass `HasFullTableCache`, but `Admin` doesn't and admin activity probably isn't a significant contributor to database traffic.

Reply:
And it's also pretty rare for admins to access any data other than their own record.